### PR TITLE
[special] Add ndtr

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #50345

Plot: 
image
https://colab.research.google.com/drive/1Of67A042rOImj8wrLF_fUTgoy_wVEOZS?usp=sharing

TODO:
* [x] Add docs (https://13385714-65600975-gh.circle-artifacts.com/0/docs/special.html#torch.special.ndtr)

### CODE REVIEW

File: test/test_unary_ufuncs.py [hunk 1]
```diff
@@ -1219,6 +1219,26 @@ def check_equal(t):
...
+    @dtypes(torch.float32, torch.float64)
+    @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
+    def test_special_ndtr_vs_scipy(self, device, dtype):
+        def check_equal(t):
+            # Test by comparing to scipy
+            actual = torch.special.ndtr(t)
+            expected = scipy.special.ndtr(t.cpu().numpy())
+            self.assertEqual(actual, expected)
+
+        range = (-10, 10)
```
Comment (kshitij12345) [inline test/test_unary_ufuncs.py:1231]: Added the test to verify the output on granular scale between this range (based on plot)

File: aten/src/ATen/native/Math.h [hunk 2]
```diff
@@ -1291,3 +1291,27 @@ calc_i0e(T _x) {
...
+  const T z = std::abs(x_mul_SQRTH);
+
+  if (z < SQRTH) {
+    return (HALF + HALF * std::erf(x_mul_SQRTH));
+  }
+  if (x_mul_SQRTH > 0) {
+    return 1.0 - (HALF * std::erfc(z));
+  }
+  return HALF * std::erfc(z);
+}
```
Comment (kshitij12345) [inline aten/src/ATen/native/Math.h:1317]: Boost has similar implementation : https://www.boost.org/doc/libs/1_76_0/boost/math/distributions/normal.hpp

Comment (mruberry) [reply aten/src/ATen/native/Math.h:1317]: How easy is it to review the numeric accuracy of a composite implementation?

File: test/test_unary_ufuncs.py [hunk 3]
```diff
@@ -1219,6 +1219,26 @@ def check_equal(t):
...
+
+        t = torch.linspace(*range, int(1e4), device=device, dtype=dtype)
+        check_equal(t)
+
+        # NaN, inf, -inf are tested in reference_numerics tests.
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:1236]: Aren't they tested in reference_numerics_extremal?

File: test/test_unary_ufuncs.py [hunk 4]
```diff
@@ -1219,6 +1219,26 @@ def check_equal(t):
         t = torch.tensor([min, max, eps, tiny], dtype=dtype, device=device)
         check_equal(t)
 
+    @dtypes(torch.float32, torch.float64)
+    @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
+    def test_special_ndtr_vs_scipy(self, device, dtype):
```
Comment (mruberry) [inline test/test_unary_ufuncs.py:1224]: Do we really need this test?

**Review by mruberry — APPROVED** [review]
This looks great!

What's your thinking on the specialized unary ufunc test? Maybe it's time to redo the reference numerics tests...

File: test/test_unary_ufuncs.py (hunk 3 above)
Comment (kshitij12345) [reply test/test_unary_ufuncs.py:1236]: What I meant was that these values are already tested to so we don't need to test them. Instead we test other extremal values like `eps`, `min`, `max`, `tiny`.
Should I update the comment in case we do keep the test.

File: test/test_unary_ufuncs.py (hunk 4 above)
Comment (kshitij12345) [reply test/test_unary_ufuncs.py:1224]: I think it is okay to have this test for now.

I did try previously to add `torch.finfo(eps), torch.finfo(max), etc` to reference numerics but few tests failed.

So, as a follow-up to `i1`, `i1e` and this PR, I will update the reference numerics (after syncing with @pmeier) to extend the range of `make_tensor` and also add `eps, tiny, min, max` from finfo to `reference_numerics_extreme` .

Previous conversation: https://github.com/pytorch/pytorch/pull/56352#discussion_r633208717

**Comment by mruberry** [comment]
Unfortunately this picked up a merge conflict, @kshitij12345

**Comment by anjali411** [comment]
@kshitij12345 why doesn't this function have an entry in `derivatives.yaml`?
